File: java/jimmer-simple/src/main/java/org/babyfish/jimmer/sql/example/web/EntityStyleController.java
package org.babyfish.jimmer.sql.example.web;

import org.babyfish.jimmer.client.FetchBy;
import org.babyfish.jimmer.sql.JSqlClient;
import org.babyfish.jimmer.sql.example.model.*;
import org.babyfish.jimmer.sql.fetcher.Fetcher;
import org.jetbrains.annotations.Nullable;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import java.util.List;

@RestController
@RequestMapping("/entity")
@Transactional
public class EntityStyleController implements Tables, Fetchers {

    private final JSqlClient sqlClient;

    public EntityStyleController(JSqlClient sqlClient) {
        this.sqlClient = sqlClient;
    }

    /**
     * Find simple book objects.
     *
     * <p>If this parameter is specified, only books with the specified name will be queried.</p>
     *
     * @param name Optional query parameter,
     *      {@code null} or {@code ""} will be ignored
     * @return A list of Book objects, each of which
     *      has only {@code id} and {@code name} properties
     */
    @GetMapping("/books/simple")
    public List<@FetchBy("SIMPLE_BOOK") Book> findSimpleBooks(
            @Nullable @RequestParam(required = false) String name
    ) {
        BookTable table = BOOK_TABLE;
        return sqlClient.createQuery(table)
                .where(table.name().eqIf(name))
                .select(table.fetch(SIMPLE_BOOK))
                .execute();
    }

    /**
     * Find complex book objects.
     *
     * <p>If this parameter is specified, only books with the specified name will be queried.</p>
     *
     * @param name Optional query parameter,
     *      {@code null} or {@code ""} will be ignored
     * @return A list of Book objects, each of which
     *      has all scalar properties, associated
     *      {@link BookStore} and associated
     *      {@link Author} objects.
     */
    @GetMapping("/books/complex")
    public List<@FetchBy("COMPLEX_BOOK") Book> findComplexBooks(
            @Nullable @RequestParam(required = false) String name
    ) {
        BookTable table = BOOK_TABLE;
        return sqlClient.createQuery(table)
                .where(table.name().eqIf(name))
                .select(table.fetch(COMPLEX_BOOK))
                .execute();
    }

    /**
     * Query root tree nodes, that is, nodes whose parentId is null.
     *
     * @return All list of root tree nodes, each of which
     *         has only {@code id} and {@code name} properties
     */
    @GetMapping("/rootTreeNodes/simple")
    public List<@FetchBy("SIMPLE_TREE_NODE") TreeNode> findSimpleTreeNodes() {
        TreeNodeTable table = TREE_NODE_TABLE;
        return sqlClient.createQuery(table)
                .where(table.parentId().isNull())
                .select(table.fetch(SIMPLE_TREE_NODE))
                .execute();
    }

    /**
     * Query root tree nodes, that is, nodes whose parentId is null.
     *
     * @return All list of root tree nodes, each of which
     *         has all scalar properties, associated child objects.
     *         If a child node has deeper child nodes,
     *         it will be recursively associated until
     *         there are no deeper child nodes.
     */
    @GetMapping("/rootTreeNodes/recursive")
    public List<@FetchBy("RECURSIVE_TREE_NODE") TreeNode> findRecursiveTreeNodes() {
        TreeNodeTable table = TREE_NODE_TABLE;
        return sqlClient.createQuery(table)
                .where(table.parentId().isNull())
                .select(table.fetch(RECURSIVE_TREE_NODE))
                .execute();
    }

    /**
     * Simple Book DTO defined by ObjectFetcher
     */
    private static final Fetcher<Book> SIMPLE_BOOK =
            BOOK_FETCHER
                    .name();

    /**
     * Complex Book DTO defined by ObjectFetcher
     */
    private static final Fetcher<Book> COMPLEX_BOOK =
            BOOK_FETCHER
                    .allScalarFields()
                    .store(
                            BOOK_STORE_FETCHER
                                    .allScalarFields()
                    )
                    .authors(
                            AUTHOR_FETCHER
                                    .allScalarFields()
                    );

    /**
     * Simple TreeNode DTO defined by ObjectFetcher
     */
    private static final Fetcher<TreeNode> SIMPLE_TREE_NODE =
            TREE_NODE_FETCHER
                    .name();

    /**
     * Recursive TreeNode defined by ObjectFetcher
     */
    private static final Fetcher<TreeNode> RECURSIVE_TREE_NODE =
            TREE_NODE_FETCHER
                    .allScalarFields()
                    .recursiveChildNodes();
}

File: kotlin/jimmer-simple-kt/src/main/kotlin/org/babyfish/jimmer/sql/example/kt/web/EntityStyleController.kt
package org.babyfish.jimmer.sql.example.kt.web

import org.babyfish.jimmer.client.FetchBy
import org.babyfish.jimmer.sql.example.kt.model.*
import org.babyfish.jimmer.sql.kt.KSqlClient
import org.babyfish.jimmer.sql.kt.ast.expression.`eq?`
import org.babyfish.jimmer.sql.kt.ast.expression.isNull
import org.babyfish.jimmer.sql.kt.fetcher.newFetcher
import org.springframework.transaction.annotation.Transactional
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.RestController

@RestController
@RequestMapping("/entity")
@Transactional
class EntityStyleController(
    private val sqlClient: KSqlClient
) {

    /**
     * Find simple book objects.
     *
     * If this parameter is specified, only books with the specified name will be queried.
     *
     * @param name Optional query parameter,
     * `null` and `""` will be ignored
     *
     * @return A list of Book objects, each of which
     * has only `name` and `edition` properties
     */
    @GetMapping("/books/simple")
    fun findSimpleBooks(
        @RequestParam(required = false) name: String?
    ): List<@FetchBy("SIMPLE_BOOK") Book> =
        sqlClient.executeQuery(Book::class) {
            where(table.name `eq?` name)
            select(table.fetch(SIMPLE_BOOK))
        }

    /**
     * Find complex book objects.
     *
     * If this parameter is specified, only books with the specified name will be queried.
     *
     * @param name Optional query parameter,
     * `null` and `""` will be ignored
     *
     * @return A list of Book objects, each of which
     * has all scalar properties, associated
     * [BookStore] and associated
     * [Author] objects.
     */
    @GetMapping("/books/complex")
    fun findComplexBooks(
        @RequestParam(required = false) name: String?
    ): List<@FetchBy("COMPLEX_BOOK") Book> =
        sqlClient.executeQuery(Book::class) {
            where(table.name `eq?` name)
            select(table.fetch(COMPLEX_BOOK))
        }

    /**
     * Query root tree nodes, that is, nodes whose parentId is null.
     *
     * @return All list of root tree nodes, each of which
     * has only `id` and `name` properties
     */
    @GetMapping("/rootTreeNodes/simple")
    fun findSimpleTreeNodes(
    ): List<@FetchBy("SIMPLE_TREE_NODE") TreeNode> =
        sqlClient.executeQuery(TreeNode::class) {
            where(table.parentId.isNull())
            select(table.fetch(SIMPLE_TREE_NODE))
        }

    /**
     * Query root tree nodes, that is, nodes whose parentId is null.
     *
     * @return All list of root tree nodes, each of which
     *         has all scalar properties, associated child objects.
     *         If a child node has deeper child nodes,
     *         it will be recursively associated until
     *         there are no deeper child nodes.
     */
    @GetMapping("/rootTreeNodes/recursive")
    fun findRecursiveTreeNodes(
    ): List<@FetchBy("RECURSIVE_TREE_NODE") TreeNode> =
        sqlClient.executeQuery(TreeNode::class) {
            where(table.parentId.isNull())
            select(table.fetch(RECURSIVE_TREE_NODE))
        }

    companion object {

        /**
         * Simple Book DTO defined by ObjectFetcher
         */
        private val SIMPLE_BOOK =
            newFetcher(Book::class).by {
                name()
            }

        /**
         * Complex Book DTO defined by ObjectFetcher
         */
        private val COMPLEX_BOOK =
            newFetcher(Book::class).by {
                allScalarFields()
                store {
                    allScalarFields()
                }
                authors {
                    allScalarFields()
                }
            }

        /**
         * Simple TreeNode DTO defined by ObjectFetcher
         */
        private val SIMPLE_TREE_NODE =
            newFetcher(TreeNode::class).by {
                name()
            }

        /**
         * Recursive TreeNode DTO defined by ObjectFetcher
         */
        private val RECURSIVE_TREE_NODE =
            newFetcher(TreeNode::class).by {
                allScalarFields()
                `childNodes*`()
            }
    }
}